bindings/time-profiler.cc

```diff
@@ -0,0 +1,219 @@
+/**
+ * Copyright 2021 Datadog Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <sstream>
+#include <vector>
+#include <map>
+
+#include "time-profiler.h"
+
+// TODO(qard): Add source map support
+TimeProfileEncoder::TimeProfileEncoder(const Napi::Env& env,
+                                       v8::CpuProfiler* profiler,
+                                       v8::Local<v8::String> name,
+                                       int nanos)
+  : PromiseWorker(env),
+    cpuProfiler(profiler),
+    cpuProfile(cpuProfiler->StopProfiling(name)),
+    intervalNanos(nanos),
+    startTime(now()) {}
+
+TimeProfileEncoder::~TimeProfileEncoder() {
+  cpuProfile->Delete();
+  cpuProfile = nullptr;
+
+  cpuProfiler->Dispose();
+  cpuProfiler = nullptr;
+}
+
+void TimeProfileEncoder::Execute() {
+  pprof::ValueType sample_count("sample", "count");
+  pprof::ValueType wall_nanoseconds("wall", "nanoseconds");
+  pprof::Profile profile(sample_count, wall_nanoseconds);
+  profile.time_nanos = startTime;
+  profile.period = intervalNanos;
+
+  // Duration nanos
+  auto startTime = cpuProfile->GetStartTime();
+  auto endTime = cpuProfile->GetEndTime();
+  profile.duration_nanos = (endTime - startTime) * 1000;
+
+  // Add root children to queue with empty stacks
+  std::map<const v8::CpuProfileNode*, std::vector<pprof::Location>> nodeStacks;
+  std::vector<const v8::CpuProfileNode*> children;
+
+  auto root = cpuProfile->GetTopDownRoot();
+  for (int i = 0; i < root->GetChildrenCount(); i++) {
+    auto child = root->GetChild(i);
+    nodeStacks[child] = {};
+    children.push_back(child);
+  }
+
+  // Process node queue
+  while (children.size() > 0) {
+    // Get last node
+    auto node = children.back();
+    children.pop_back();
+
+    // Find function and script names
+    auto scriptName = fallback(node->GetScriptResourceNameStr(), "<native>");
+    auto name = fallback(node->GetFunctionNameStr(), "(anonymous)");
```

Comment:
maybe add `<native>` and `(anonymous)` in static field (ex `static const std::string native_string = "<native>";` )
This will prevent creating a `std::string` at each call.